docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,294 @@
+
+The encoding layout of TiDB has been described in our [previous article](https://pingcap.com/blog/2017-07-11-tidbinternal2/#map). The row format should be changed to make it memory comparable, this is important to the index lookup. Basic principle is that all keys encoded for strings should use the `sortKeys` result from `Key()`/`KeyFromString()` function. However, most of the `sortKeys` calculations are not reversible.
+
+#### Option 1
+
+  * For table data, encodings stay unchanged. All strings are compared after decoding with the `Compare()` function.
+  * For table indices, we replace current `ColumnValue` with `sortKey`:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID
+	```
+    - For non-unique indices:
+    ```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: null
+	```
+
+Pros:
+    The current encoding layout keeps unchanged.
+
+Cons:
+    Since index value can not be recovered from `sortKey`, extra table lookup is needed for current index-only queries, this may lead to performance regressions and the "covering index" can not work. The extra table lookup also requires corresponding changes to the TiDB optimizer.
+
+#### Option 2
+
+  * Table data encodings is the same with Option 1.
+  * For table indices, we encode the `ColumnValue` to the value of the storage, so as to avoid extra table lookup for index queries:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID_indexedColumnsValue
+	```
+    - For non-unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: indexedColumnsValue
+	```
+
+Pros: Extra table lookup is avoided and no change is needed to the TIDB optimizer. What's more, it is able to scan just the index in order to upgrade the on-disk storage to fix bugs in collations, because it has the `sortKey` and the value in the index.
+
+Cons: The size of index value on string column with new collations is doubled(we need to write both `sortKey` and `indexedColumnsValue`, their sizes are equal for most of the collations).
+
+## Compatibility
+
+### Compatibility with MySQL
+
+As stated before, MySQL has various different collation implementations. For example, in MySQL 5.7, the default collation of `utf8mb4` is `utf8mb4_general_ci`, which has two main flaws:
+
+  * It only implement the `simple mappings` of UCA.
+  * It used a very old version of UCA.
+
+In MySQL 8, the default collation of `utf8mb4` has been changed to `utf8mb4_0900_ai_ci`, in which the Unicode version is upgraded to 9.0, with full UCA implementation.
+
+Should TiDB aim to support all collation of MySQL?
+
+  * If TiDB supports the latest collations only(for example, support `utf8mb4_0900_ai_ci` only and treat `utf8mb4_general_ci` as one of its alias). This seems to be 'standard-correct' but may lead to potential compatibility issues.
```

Comment:
Do we have decisions (which option is preferred) now?

Reply:
Not yet